File: articles/mathematical-biology-logic-living-systems/fortran/math_biology_kernel.f90
! Compact mathematical-biology numerical kernel in Fortran.

program math_biology_kernel
  implicit none

  integer :: step, nsteps
  real :: t, dt
  real :: n0, r, k, logistic_final
  real :: s, i, recovered, beta, gamma, ds, di, dr
  real :: peak_i, time_to_peak
  real :: prey, predator, alpha, pred_beta, delta, pred_gamma
  real :: dprey, dpredator
  real :: substrate, vmax, km, velocity

  n0 = 100.0
  r = 0.30
  k = 2000.0
  t = 40.0
  logistic_final = k / (1.0 + ((k - n0) / n0) * exp(-r * t))

  beta = 0.35
  gamma = 0.10
  s = 0.99
  i = 0.01
  recovered = 0.0
  dt = 0.05
  nsteps = int(120.0 / dt)
  peak_i = i
  time_to_peak = 0.0

  do step = 1, nsteps
    if (i > peak_i) then
      peak_i = i
      time_to_peak = step * dt
    end if

    ds = -beta * s * i
    di = beta * s * i - gamma * i
    dr = gamma * i

    s = max(s + ds * dt, 0.0)
    i = max(i + di * dt, 0.0)
    recovered = max(recovered + dr * dt, 0.0)
  end do

  prey = 40.0
  predator = 9.0
  alpha = 0.60
  pred_beta = 0.025
  delta = 0.018
  pred_gamma = 0.35
  dt = 0.01
  nsteps = int(80.0 / dt)

  do step = 1, nsteps
    dprey = alpha * prey - pred_beta * prey * predator
    dpredator = delta * prey * predator - pred_gamma * predator

    prey = max(prey + dprey * dt, 0.0)
    predator = max(predator + dpredator * dt, 0.0)
  end do

  substrate = 5.0
  vmax = 10.0
  km = 2.0
  velocity = vmax * substrate / (km + substrate)

  print *, "Logistic final:", logistic_final
  print *, "SIR peak infected:", peak_i
  print *, "SIR time to peak:", time_to_peak
  print *, "SIR final recovered:", recovered
  print *, "Final prey:", prey
  print *, "Final predator:", predator
  print *, "Michaelis-Menten velocity:", velocity
end program math_biology_kernel

File: articles/mathematical-biology-logic-living-systems/c/math_biology_kernel.c
/*
 * Compact mathematical-biology numerical kernel in C.
 */

#include <math.h>
#include <stdio.h>

double logistic_growth(double t, double n0, double r, double k) {
    return k / (1.0 + ((k - n0) / n0) * exp(-r * t));
}

double michaelis_menten(double substrate, double vmax, double km) {
    return vmax * substrate / (km + substrate);
}

int main(void) {
    double logistic_final = logistic_growth(40.0, 100.0, 0.30, 2000.0);
    double enzyme_velocity = michaelis_menten(5.0, 10.0, 2.0);

    double beta = 0.35;
    double gamma = 0.10;
    double dt = 0.05;
    int steps = (int)(120.0 / dt);

    double s = 0.99;
    double i = 0.01;
    double r = 0.0;
    double peak_i = i;
    double time_to_peak = 0.0;

    for (int step = 0; step < steps; step++) {
        double time = step * dt;

        if (i > peak_i) {
            peak_i = i;
            time_to_peak = time;
        }

        double ds = -beta * s * i;
        double di = beta * s * i - gamma * i;
        double dr = gamma * i;

        s = fmax(s + ds * dt, 0.0);
        i = fmax(i + di * dt, 0.0);
        r = fmax(r + dr * dt, 0.0);
    }

    printf("logistic_final=%.6f\n", logistic_final);
    printf("michaelis_menten_velocity=%.6f\n", enzyme_velocity);
    printf("sir_peak_infected=%.6f\n", peak_i);
    printf("sir_time_to_peak=%.6f\n", time_to_peak);
    printf("sir_final_recovered=%.6f\n", r);

    return 0;
}
